C# 7.x: ref struct

gafter opened the pull request:
Split out `ref struct` from #333.

This still needs to safety rules to be incorporated.  See https://github.com/dotnet/csharpstandard/issues/334 and https://github.com/dotnet/csharplang/blob/standard-proposals/proposals/csharp-7.2/span-safety.md#draft-language-specification .  Those safety rules are for ref return, ref locals, and ref structs.

Fixes #267 
Fixes #334

Commit:
Split out `ref struct` from #33.

KalleOlaviNiemitalo commented on standard/structs.md:
```diff
@@ -49,6 +51,22 @@ It is a compile-time error for the same modifier to appear multiple times in a s
 
 The modifiers of a struct declaration have the same meaning as those of a class declaration ([§14.2.2](classes.md#1422-class-modifiers)).
 
+### §ref-modifier-new-clause Ref modifier
+
+The `ref` modifier indicates that the *struct_declaration* declares a type whose instances are allocated on the execution stack.
+
+It is a compile-time error if a ref struct type is used in any of the following contexts:
+
+- As the element type of an array.
+- As the declared type of a field of a class or a non-ref struct.
+- To implement an interface.
+- Being boxed to `System.ValueType` or `System.Object`.
+- As a type argument.
+- An async method.
+- An iterator.
```
Does "a ref struct type is used" include this case:

```C#
using System;
using System.Threading.Tasks;

class Widget
{
    public static implicit operator Widget(Span<int> s) => null;
}

static class Class1
{
    static async Task M(Task t)
    {
        Widget w = stackalloc int[] { 42 };
        await t;
    }
}
```

.NET SDK 6.0.101 compiles it without errors. The Span\<int\> becomes a local variable in the MoveNext method of the state machine.

__Originally posted by @KalleOlaviNiemitalo in <https://github.com/dotnet/csharpstandard/pull/333#discussion_r775953793>__

gafter replied:
Yes.  The implementation is more permissive than required.

KalleOlaviNiemitalo replied:
Ah, so "It is a compile-time error" does not mean "The compiler shall issue an error".

jskeet replied:
Action from 2023-04-19: we should just keep a note of this in the non-conformance doc (which is not yet merged/decided on). No change required to this PR.

BillWagner replied:
See dotnet/docs#35096 to capture this issue.

BillWagner commented:
Rex had written several questions in #334  Reviewers should check that those are addressed.

BillWagner commented:
@gafter 

Do you want me to pick up this one? I'm working on updating the language for #213, #219, #238, and possibly #240 for the next meeting. I'll also be adding the sections on ref safety.

If so, I can add it to my list and get this one ready as well.

gafter commented:
@BillWagner Yes, please.  If you do, please try to address #267 as well.

BillWagner committed:
Add ref safety rules as they apply to ref structs

Declare that `ref structs` have a *ref_safe_scope* that matches their initializing expressions. Restrict copying of a `ref struct` (by value) to its *ref_safe_scope*. Then, define what the *ref_safe_scope* is depending on the initializing expression.

BillWagner commented:
@jskeet @gafter 

I added a new clause to define the ref safety rules as they apply to ref structs. This should be ready for our next meeting.

BillWagner committed:
forgot to finish one sentence.

BillWagner commented:
> @BillWagner Yes, please. If you do, please try to address #267 as well.

The normative language was already present. I added a note to make it more obvious.

BillWagner committed:
Add note on iterators and async methods

The previous normative language was

jskeet commented on standard/structs.md:
```diff
@@ -16,18 +16,20 @@ A *struct_declaration* is a *type_declaration* ([§14.7](namespaces.md#147-type-
 
 A struct declaration that supplies a *type_parameter_list* is a generic struct declaration. Additionally, any struct nested inside a generic class declaration or a generic struct declaration is itself a generic struct declaration, since type arguments for the containing type shall be supplied to create a constructed type ([§8.4](types.md#84-constructed-types)).
 
+A struct declaration that includes a `ref` keyword may not have a *struct_interfaces* part.
```
Shall not rather than may not, for clarity? One for Rex.

jskeet replied:
Action from 2023-04-19: let's use shall not or "it is an error to" etc.

jskeet commented on standard/structs.md:
```diff
@@ -60,6 +62,26 @@ A readonly struct has the following constraints:
 
+### §ref-modifier-new-clause Ref modifier
+
+The `ref` modifier indicates that the *struct_declaration* declares a type whose instances are allocated on the execution stack. The `ref` modifier declares that instances may contain ref-like fields, and must obey the *ref_safe_scope* rules. Instances of the ref struct may not be copied out of the *ref_safe_scope* of any of its ref-like fields. The rules for determining the *safe_scope* of a `re ...
```
This has both ref safe scope and safe scope... Needs slight rework with the recent changes?

jskeet commented on standard/structs.md:
```diff
@@ -60,6 +62,26 @@ A readonly struct has the following constraints:
+- An async method.
+- An iterator.
+
+> *Note*: A `ref struct` may not declare `async` methods nor use a `yield return` or `yield break` statement because the implicit `this` parameter cannot be used in those contexts. *end note*
```
More "may not" that should be shall not, I think.

BillWagner replied:
Pint @RexJaeschke again.

jskeet replied:
Action from 2023-04-19: As before, @BillWagner to change to anything unambiguous.

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +498,19 @@ Static constructors for structs follow most of the same rules as for classes. Th
+- If `e` designates a reference to a conditional, `ref c ? ref e1 : ref e2`, the *ref_safe_scope* of `S` is the narrowest scope among the *ref_safe_scopes* of `e1` and `e2`.
+- If `e` is the result of a method invocation, `e1.M(e2, ...)`, the *ref_safe_scope* of `S` is the narrowest of the *ref_safe_scope*s of all argument expressions, including the receiver.
+
+For an assignment `e1 = e2`, if the type of `e1` is a `ref struct` type, then the *ref_safe_scope* of `e1` is the *ref_safe_scope* of `e2`.
```
Not sure how assignment matters here. Would like to hear more in the meeting.

BillWagner replied:
Using a modified version of another example:

```csharp
    ReadOnlySpan<char> span1;
    ReadOnlySpan<char> span3;
    if (x is not null)
    {
        ReadOnlySpan<char> span2 = x.AsSpan(); // span2 ref_safe_scope is *method*.
        span1 = span2;     // span1 ref_safe_scope is *method* (same as span2)
        span3 = span1; // span3 ref_safe_scope is method (same as span1)
    }
}
```

jskeet replied:
Discussion in meeting: a *variable* `v` has a *ref_safe_scope*, and if `v` can be assigned variable references which have different *ref_safe_scopes*, it would be odd for `v`'s *ref_safe_scope* to change.

As per https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md#language-constraints, the assignment part should be a constraint on what *can* be assigned, but a determination of the *ref_safe_scope* of the LHS.

jskeet replied:
Also noted from @gafter: this would be safe_to_escape not ref_safe_to_escape in the previous way of explaining things.

Action 2023-04-19: @BillWagner to consider - and ping others where appropriate.

gafter replied:
These should be language constraints, as per https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md#language-constraints, not inference rules.

BillWagner replied:
Update per meeting notes.

jskeet reviewed (approved):
I think this is basically fine, with a couple of notes. I may well be missing subtleties though.

Nigel-Ecma commented on standard/structs.md:
```diff
@@ -476,3 +498,19 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+A `ref struct` may contain *ref_like field*s. A *ref_like field* is a *reference_variable* (§ref-span-safety). An instance of a `ref struct` may not be copied outside the *ref_safe_scope* (§ref-span-safety-escape-scopes) of any of its ref-like fields. The *ref_safe_scope* of a ref struct is determined from its initializing expression.
+
+The default value of ref-like fields is the value of the `default` expression, a null reference. The *ref_safe_scope* of a null reference is *caller_scope*. A `ref struct` initialized to its default value can be copied to the *caller_scope*.
```
“The default value of ref-like fields is the value of the `default` expression, a null reference.” does not read well. Is “…, which is a null reference" the intention? Or is it “The default value of ref-like fields, and the value of the `default` expression, is a null reference.”? Should there be an xref to the default literal/operator? Sorry, only questions not answers!

BillWagner committed:
respond to review feedback.

BillWagner commented on standard/structs.md:
```diff
@@ -60,6 +62,26 @@ A readonly struct has the following constraints:
+- As the element type of an array.
+- As the declared type of a field of a class or a non-ref struct.
+- To implement an interface.
+- Being boxed to `System.ValueType` or `System.Object`.
```
Should this include `dynamic` as well?

jskeet reviewed (commented):
Result from meeting: this is definitely not right at the moment, but we're not sure how hard it will be to fix. @BillWagner has the lead.

BillWagner committed:
Respond to meeting feedback, part 1

Respond to all meeting feedback *except* updating the rules on *safe_to_escape*.

That's coming in the next commit.

BillWagner committed:
incorporate safe rules.

Pull all safety rules related to safe-scope from PR on ref variables
into the section on ref structs.

RexJaeschke commented:
@BillWagner In #### safe-scope-rules-general General

> At compile-time, each expression whose type is a ref struct is associated with a scope where that instance is safe, its ***safe-scope***. The ***safe-scope*** is a scope, enclosing an expression, to which it is safe for the value to escape to. If that scope is the entire function member, we say that the value is ***safe-to-return*** from the function member.

the term ***safe-scope*** is defined twice w.r.t term formatting. Remove the formatting from the second occurrence.

BillWagner commented:
> @BillWagner In #### safe-scope-rules-general General
> 
> > At compile-time, each expression whose type is a ref struct is associated with a scope where that instance is safe, its _**safe-scope**_. The _**safe-scope**_ is a scope, enclosing an expression, to which it is safe for the value to escape to. If that scope is the entire function member, we say that the value is _**safe-to-return**_ from the function member.
> 
> the term _**safe-scope**_ is defined twice w.r.t term formatting. Remove the formatting from the second occurrence.

Thanks Rex. Updated.

BillWagner committed:
respond to feedback.

jskeet commented on standard/structs.md:
```diff
@@ -60,6 +62,29 @@ A readonly struct has the following constraints:
+- An async method.
+- An iterator.
+
+> *Note*: A `ref struct` shall not declare `async` methods nor use a `yield return` or `yield break` statement because the implicit `this` parameter cannot be used in those contexts. *end note*
```
```suggestion
> *Note*: A `ref struct` shall not declare `async` instance methods nor use a `yield return` or `yield break` statement within an instance method, because the implicit `this` parameter cannot be used in those contexts. *end note*
```

(Or something similar - basically, static iterator blocks and static async methods are fine.)

jskeet commented on standard/structs.md:
```diff
@@ -60,6 +62,29 @@ A readonly struct has the following constraints:
+> *Note*: A `ref struct` shall not declare `async` methods nor use a `yield return` or `yield break` statement because the implicit `this` parameter cannot be used in those contexts. *end note*
+
+- There is no conversion from a `ref struct` type to the type `object` or the type `System.ValueType`.
+- A `ref struct` type shall not be declared to implement any interface
```
```suggestion
- A `ref struct` type shall not be declared to implement any interface.
```

jskeet commented on standard/structs.md:
```diff
@@ -60,6 +62,29 @@ A readonly struct has the following constraints:
+- There is no conversion from a `ref struct` type to the type `object` or the type `System.ValueType`.
+- A `ref struct` type shall not be declared to implement any interface
+- An instance method declared in `object` or in `System.ValueType` but not overridden in a `ref struct` type shall not be called with a receiver of that `ref struct` type.
+- An instance method of a `ref struct` type shall not be captured by method conversion to a delegate type.
```
```suggestion
- An instance method of a `ref struct` type shall not be captured by method group conversion to a delegate type.
```

(Maybe?)

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+#### safe-scope-rules-general General
+
+At compile-time, each expression whose type is a ref struct is associated with a scope where that instance is safe, its ***safe-scope***. The safe-scope is a scope, enclosing an expression, to which it is safe for the value to escape to. If that scope is the entire function member, we say that the value is ***safe-to-return*** from the function member.
```
Not sure what "where that instance is safe" means. Safe to use? Perhaps "where that instance can be used safely"?

jskeet replied:
```suggestion
At compile-time, each expression whose type is a ref struct is associated with a scope where that instance is safe, its ***safe-scope***. The safe-scope is a scope, enclosing an expression, which it is safe for the value to escape to. If that scope is the entire function member, we say that the value is ***safe-to-return*** from the function member.
```

(Or remove the "to" after "escape" - but we don't need or want both.)

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+The safe-scope records which scope a ref struct may be copied into. Given an assignment from an expression `E1` with a safe-scope `S1`, to an expression `E2` with safe-scope `S2`, it is an error if `S2` is a wider scope than `S1`.
+
+There are three different safe-scope values, the same as the ref-safe-scope values defined for reference variables (§ref-span-safety-escape-scopes): block, function-member, and calling-method. An expression `e1` of a ref struct type is constrained by its safe-scope as follows:
```
```suggestion
There are three different safe-scope values, the same as the ref-safe-scope values defined for reference variables (§ref-span-safety-escape-scopes): block, function-member, and caller-scope. An expression `e1` of a ref struct type is constrained by its safe-scope as follows:
```

(And the same below.)

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+- For a return statement `return e1`, the safe-scope of `e1` must be calling-method.
+- For an assignment `e1 = e2` the safe-scope of `e2` must be at least as wide a scope as the safe-scope of `e1`.
+
+For a method invocation if there is a `ref` or `out` argument of a `ref struct` type (including the receiver unless the type is `readonly`), with safe-scope `E1`, then no argument (including the receiver) may have a narrower safe-scope than `E1`.
```
```suggestion
For a method invocation if there is a `ref` or `out` argument of a `ref struct` type (including the receiver unless the type is `readonly`), with safe-scope `S1`, then no argument (including the receiver) may have a narrower safe-scope than `S1`.
```

(Just because we normally use E for the expression, and S for its scope.)

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+A local variable of a ref struct type has a safe-scope as follows:
+
+- If the variable is an iteration variable of a `foreach` loop, then the variable's safe-scope is the same as the safe-scope of the `foreach` loop's expression.
+- A local of `ref struct` type uninitialized at the point of declaration has a safe-scope of calling method.
```
```suggestion
- A local of `ref struct` type which is uninitialized at the point of declaration has a safe-scope of calling method.
```

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+#### safe-scope-rules-field Field safe scope
+
+A a reference to a field, `e.F` where `F` is a ref struct type, has a safe-scope that is the same as the safe-scope of `e`.
```
```suggestion
A reference to a field `e.F`, where the type of `F` is a ref struct type, has a safe-scope that is the same as the safe-scope of `e`.
```

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+The application of a user-defined operator is treated as a method invocation (§safe-scope-method-invocation).
+
+For an operator that yields a value, such as `e1 + e2` or `c ? e1 : e2`, the safe-scope of the result is the narrowest scope among the safe-scopes of the operands of the operator. As a consequence, for a unary operator that yields a value, such as `+e`, the safe-scope of the result is the safe-scope of the operand.
```
Is that the case for the conditional operator? I wouldn't expect the first operand to be relevant there, but I could be wrong.

BillWagner replied:
it doesn't matter, but it doesn't hurt either. The first operand must be a `bool`. Its safe scope is calling-method.  I couldn't wordsmith language that read reasonably well. Perhaps a note on the conditional operator's first operand?

Nigel-Ecma replied:
It is the safe-scope *of the result* so the first operand (`c`) is not relevant *per se* as the safe-scope is determined statically and not based on which alternative is selected at runtime. I wouldn't wordsmith this into the normative text, but a note could be added if desired.

BillWagner replied:
I added a note, as this has come up multiple times.

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+- calling-method.
+- The safe-scope of all argument expressions (including the receiver).
+
+A property invocation (either `get` or `set`) it treated as a method invocation of the underlying method by the above rules.
```
```suggestion
A property invocation (either `get` or `set`) is treated as a method invocation of the underlying method by the above rules.
```

jskeet commented on standard/structs.md:
````diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+> *Note*: These rules rely on `Span<T>` not having a constructor of the following form:
+>
+> ```csharp
+> public Span<T>(ret T p)
````
```suggestion
> public Span<T>(ref T p)
```

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+In addition the safe-scope is no wider than the smallest of the safe-scopes of all arguments and operands of all object initializer expressions, recursively, if any initializer is present.
+
+> *Note*: These rules rely on `Span<T>` not having a constructor of the following form:
```
This is an odd note, being Span-specific. What happens if someone creates a `MySpan<T>` type with a constructor like that?

BillWagner replied:
We have to cross that bridge in C# 11. Before that, `MySpan<T>` can't have a `ref` field.

jskeet commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+A `new` expression that invokes a constructor obeys the same rules as a method invocation that is considered to return the type being constructed.
+
+In addition the safe-scope is no wider than the smallest of the safe-scopes of all arguments and operands of all object initializer expressions, recursively, if any initializer is present.
```
Can we be more specific than "no wider", e.g. just "is the smallest of the safe-scopes [...]"?

jskeet reviewed (commented):
Most of these are trivial.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <jonskeet@google.com>

BillWagner committed:
respond to feedback.

Nigel-Ecma commented on standard/structs.md:
```diff
@@ -60,6 +62,29 @@ A readonly struct has the following constraints:
+- An instance method of a `ref struct` type shall not be captured by method group conversion to a delegate type.
+- A ref struct shall not be captured by a lambda expression or a local function.
+
+These constraints ensure that no variable of `ref struct` type refers to stack memory or variables that are no longer valid.
```
I think this sentence could be parsed (and I think I read it was the wrong one):

> These constraints ensure that no variable of `ref struct` type refers to (stack memory) or (variables that are no longer valid).

or as:

> These constraints ensure that no variable of `ref struct` type refers to (stack memory or variables) that are no longer valid.

Suggest clarify.

BillWagner replied:
updated in next commit.

Nigel-Ecma commented on standard/structs.md:
```diff
@@ -60,6 +62,29 @@ A readonly struct has the following constraints:
 
+### §ref-modifier-new-clause Ref modifier
+
+The `ref` modifier indicates that the *struct_declaration* declares a type whose instances are allocated on the execution stack. The `ref` modifier declares that instances may contain ref-like fields, and may not be copied out of its safe-scope (§safe-scope-rules). The rules for determining the safe scope of a `ref struct` are described in §safe-scope-rules.
```
"execution stack", "stack memory" (used below) – need to address issued raised by @RexJaeschke in #334.

BillWagner replied:
#238 includes the changes that allow `stackalloc` in safe code. The first sentence of that new clause reads:

> A stack allocation initializer is a *local_variable_initializer* (§9.2.8) that allocates a block of memory from the call stack.

Ping @RexJaeschke Where should we add this definition? I'll add it to either this PR or #238

BillWagner replied:
Decision to add in #238.  See https://github.com/dotnet/csharpstandard/pull/238/files#r1197026105

Nigel-Ecma commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+#### safe-scope-rules-local Local variable safe scope
+
+A local variable of a ref struct type has a safe-scope as follows:
+
+- If the variable is an iteration variable of a `foreach` loop, then the variable's safe-scope is the same as the safe-scope of the `foreach` loop's expression.
+- A local of `ref struct` type which is uninitialized at the point of declaration has a safe-scope of calling method.
+- Otherwise the variable's declaration requires an initializer. The variable's safe-scope is the same as the safe-scope of its initializer.
```
There is a bit of repetition here and unusual phrasing, maybe try:

```suggestion
A local variable of a ref struct type has a safe-scope as follows:

- If the variable is an iteration variable of a `foreach` loop, then the variable's safe-scope is the same as the safe-scope of the `foreach` loop's expression.
- Otherwise if the variable's declaration has an initializer then the variable's safe-scope is the same as the safe-scope of that initializer.

```

Note: I omitted that the final option means the variable is not initialised, if that is desired the final bullet can be:

```
- Otherwise the variable is uninitialized at the point of declaration and has a safe-scope of the calling method.
```

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Nigel-Ecma <perryresearch@zoot.net.nz>

BillWagner committed:
respond to feedback.

MadsTorgersen commented on standard/structs.md:
```diff
@@ -60,6 +62,29 @@ A readonly struct has the following constraints:
 
+### §ref-modifier-new-clause Ref modifier
+
+The `ref` modifier indicates that the *struct_declaration* declares a type whose instances are allocated on the execution stack. The `ref` modifier declares that instances may contain ref-like fields, and may not be copied out of its safe-scope (§safe-scope-rules). The rules for determining the safe scope of a `ref struct` are described in §safe-scope-rules.
```
Should "ref struct" be introduced as a term?

BillWagner commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
 
 > *Note*: This access restriction means that constructors in structs containing automatically implemented properties often need an explicit constructor initializer where they would not otherwise need one, to satisfy the requirement of all fields being definitely assigned before any function member is invoked or the constructor returns. *end note*
+
+### §safe-scope-rules Safe scope constraint for `ref struct` types
```
Here, as in #742, change from "scope" to "context".

BillWagner commented on standard/structs.md:
```diff
@@ -476,3 +501,71 @@ Static constructors for structs follow most of the same rules as for classes. Th
+
+The safe-scope records which scope a ref struct may be copied into. Given an assignment from an expression `E1` with a safe-scope `S1`, to an expression `E2` with safe-scope `S2`, it is an error if `S2` is a wider scope than `S1`.
+
+There are three different safe-scope values, the same as the ref-safe-scope values defined for reference variables (§ref-span-safety-escape-scopes): block, function-member, and caller-scope. An expression `e1` of a ref struct type is constrained by its safe-scope as follows:
```
Associate *safe-to-return* with *caller-scope*.

BillWagner committed:
Update per 5/17 committee meeting.

BillWagner committed:
update definition

BillWagner committed:
found one more comment to address
